src/main/scala/is/hail/stats/DownsampleCombiner.scala

```diff
@@ -0,0 +1,128 @@
+package is.hail.stats
+
+import org.apache.spark.sql.Row
+import is.hail.utils._
+
+class DownsampleCombiner(nDivisions: Int) extends Serializable {
+  var binsToCoords: Map[(Int, Int), (Double, Double, Seq[String])] = Map.empty[(Int, Int), (Double, Double, Seq[String])]
+  var initFrame: IndexedSeq[(Double, Double, Seq[String])] = FastIndexedSeq.empty[(Double, Double, Seq[String])]
+  var hasPoints: Boolean = false
+
+  var left: Double = 0.0
+  var right: Double = 0.0
+  var bottom: Double = 0.0
+  var top: Double = 0.0
+
+  def collapse(binsToCoords: Map[(Int, Int), (Double, Double, Seq[String])], nDivisions: Int,
+               factor: Int, collapseX: Boolean, collapseY: Boolean): Map[(Int, Int), (Double, Double, Seq[String])] = {
+    val mappings: Map[Int, Int] =
+      Array.range(-nDivisions * math.pow(factor, 2).toInt, nDivisions * (-math.pow(factor, 2).toInt + math.pow(3, factor).toInt))
+        .zip(0.until(nDivisions).flatMap(e => Array.fill(math.pow(3, factor).toInt)(e)))
+        .toMap
+
+    binsToCoords.map { case ((i, j), (x, y, l)) => (if (collapseX) mappings(i) else i, if (collapseY) mappings(j) else j) -> (x, y, l) }
+  }
+
+  def toBinCoords(x: Double, y: Double, left: Double, right: Double, bottom: Double, top: Double, nDivisions: Int): (Int, Int) = {
+    val i = (((x - left) / (right - left)) * nDivisions).toInt
+    val j = (((y - bottom) / (top - bottom)) * nDivisions).toInt
+    (i, j)
+  }
+
+  def buildFactor(coord: Double, lower: Double, upper: Double, factor: Int = 0): Int = {
+    assert(lower < upper)
+    if (coord < lower || coord > upper)
+      buildFactor(coord, lower - (upper - lower), upper + (upper - lower), factor + 1)
+    else
+      factor
+  }
+
+  def merge(x: Double, y: Double, label: Seq[String]): DownsampleCombiner = {
```

Comment:
I think this can be improved as we discussed in person with baseMerge and a collapse function that uses the point to recompute the bins when collapsing.